packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -2986,14 +3000,17 @@ class _RouteEntry extends RouteTransitionRecord {
 
   void markNeedsExitingDecision() => _isWaitingForExitingDecision = true;
 
+  _TransitionDecision? pendingDecision;
+  Object? pendingResult;
+
   @override
   void markForPush() {
     assert(
       isWaitingForEnteringDecision && !isWaitingForExitingDecision,
       'This route cannot be marked for push. Either a decision has already been '
       'made or it does not require an explicit decision on how to transition in.',
     );
-    currentState = _RouteLifecycle.push;
+    pendingDecision = _TransitionDecision.push;
```

Comment:
Is there any reason that push/add/complete need to take the same path? Could this be simplified by only delaying pops?

Reply:
yeah, I think so, but I think that would apply to remove/complete as well since those operations will actually change the state of the route. It feels weird that entering decision and existing decisions are driven differently. for entering, we set currentState to initiate the didPush/didAdd on the route, while the exiting, we call pop/compete/remove to set the currentState. they should be unify somehow...